csi: add custom ceph.conf for csi pods (#9567)
rook/rook: Documentation/ceph-advanced-configuration.md

    osd pool default size = 2
```

## Custom CSI ceph.conf Settings

> **WARNING**: It is highly recommended to use the default setting that comes with
> CephCSI and this can only be used when absolutely necessary.

> > The `ceph.conf` be reset back to default values if/when the configurations are no

Reviewer: ```suggestion
> The `ceph.conf` should be reset back to default values if/when the configurations are no
```

    osd pool default size = 2
```

## Custom CSI ceph.conf Settings

> **WARNING**: It is highly recommended to use the default setting that comes with
> CephCSI and this can only be used when absolutely necessary.
> The `ceph.conf` be reset back to default values if/when the configurations are no
> longer necessary.

If the `csi-ceph-conf-override` ConfigMap is created before the cluster is
started, the CephCSI pods will automatically pick up the settings. If you
add the settings to the ConfigMap after the cluster has been initialized,
you can restart the Rook operator pod and wait for Rook to recreate CSI pods
to take immediate effect.

After the CSI pods are restarted, the new settings should be in effect.

### Example

In this [Example](https://github.com/rook/rook/tree/{{ branchName }}/deploy/csi-ceph-conf-override.yaml) we

> will set the `rbd_validate_pool` to `false` to skip rbd pool validation

Reviewer: ```suggestion
will set the `rbd_validate_pool` to `false` to skip rbd pool validation.
```

Reviewer: nit

> **WARNING**: Modify Ceph settings carefully to avoid modifying the default
> configuration.
> Changing the settings could result in unexpected results if used incorrectly.

```console

> cat csi-ceph-conf-override.yaml

Reviewer: I might suggest just add a link to the example yaml and not duplicate it here. But if you want to keep it here, no need for the `cat` command
```suggestion
```

Author: Removed the cat and content

apiVersion: v1
kind: ConfigMap
metadata:
  name: csi-ceph-conf-override
  namespace: rook-ceph
data:
  ceph.conf: |
    [global]
    auth_cluster_required = cephx
    auth_service_required = cephx
    auth_client_required = cephx
    rbd_validate_pool = false

>   # keyring is a required key and its value should be empty

Reviewer: Why should the keyring be empty?

Author: removed it now. Mounting configmap as subpath to csi pods.

  keyring: |
```

> **WARNING**: Don't delete/modify the keyring value it should be left empty.
> Changing keyring settings could result in unexpected results.

```console
kubectl create -f csi-ceph-conf-override.yaml
```

> Restart Rook operator pod and wait for CSI pods to recreate.

Reviewer: ```suggestion
Restart the Rook operator pod and wait for the CSI pods to be recreated.
```

validation

> **WARNING**: Modify Ceph settings carefully to avoid modifying the default
> configuration.
> Changing the settings could result in unexpected results if used incorrectly.

```console
kubectl -n rook-ceph get ConfigMap csi-ceph-conf-override -o yaml
```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: csi-ceph-conf-override
  namespace: rook-ceph
data:
  ceph.conf: |
    [global]
    auth_cluster_required = cephx
    auth_service_required = cephx
    auth_client_required = cephx

    fuse_set_user_groups = false
    fuse_big_writes = true
  # keyring is a required key and its value should be empty
  keyring: |
```

To apply your desired configuration, you will need to update this ConfigMap. The next time the
CephCSI pod(s) restarts, they will use the updated configs.

```console
kubectl -n rook-ceph edit configmap csi-ceph-conf-override
```

> **WARNING**: Don't delete/modify the keyring value it should be left empty.
> Changing keyring settings could result in unexpected results.

Modify the settings and save. Each line you add should be indented from the `ceph.conf` property as such:

> ```yaml

Reviewer: Can we keep this section simpler by having just the one example above? Showing a single line added could be done in the example above.

    osd pool default size = 2
```

## Custom CSI ceph.conf Settings

> **WARNING**: It is highly recommended to use the default setting comes with
> CephCSI and this can only be used when absolutely necessary and that the
> `ceph.conf` be reset back to default values if/when the configurations are no
> longer necessary.

> **WARNING**: Rook/CephCSI performs no validation on the config, so the validity of the settings is the
> user's responsibility.

If the `csi-ceph-conf-override` ConfigMap is created before the cluster is
started, the CephCSI pods
will automatically pick up the settings. If you add the settings to the ConfigMap after the cluster
has been initialized, you can restart the CSI pods to take immediate effect

After the CSI pods are restarted, the new settings should be in effect.

### Example

In this example we will set the `rbd_validate_pool` to `false` to skip rbd pool
validation

> **WARNING**: Modify Ceph settings carefully avoid modifying the default

> > configuration. You are leaving the sandbox tested by Rook.

Author: copy paste from Rook section removed it.

> > Changing the settings could result in unexpected results or even data loss if used incorrectly.

Author: removed copy paste mistake.

```console
kubectl -n rook-ceph get ConfigMap csi-ceph-conf-override -o yaml
```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: csi-ceph-conf-override
  namespace: rook-ceph
data:
  ceph.conf: |
    [global]
    auth_cluster_required = cephx
    auth_service_required = cephx
    auth_client_required = cephx

    fuse_set_user_groups = false
    fuse_big_writes = true
  # keyring is a required key and its value should be empty
  keyring: |
```

To apply your desired configuration, you will need to update this ConfigMap. The next time the

> CephCSI pod(s) start, they will use the updated configs.

Author: Yes, correct. even it gets auto-updated. asking to wait for restart to make sure it's updated as ceph.conf will get applied if remounts happen.
